src/Hintman/GitHub/Auth.hs

```diff
@@ -0,0 +1,84 @@
+{- | Functions for authenticating GitHub app. -}
+
+module Hintman.GitHub.Auth
+       ( JwtToken (..)
+       , GithubWebhookSecret (..)
+       , mkGitHubAuth
+       ) where
+
+import Data.Time.Clock.POSIX (POSIXTime, getPOSIXTime)
+import System.Environment (getEnv)
+
+import qualified Web.JWT as Jwt
+
+
+{- | Notice that the private key must be in PEM format, but the newlines should
+be stripped and replaced with the literal `\n`. This can be done in the terminal
+as such:
+
+@
+export GITHUB_PRIVATE_KEY=`awk '{printf "%s\\n", $0}' private-key.pem`
+@
+
+Env variable: GITHUB_PRIVATE_KEY
+-}
+newtype GithubPrivateKey = GithubPrivateKey Text
+
+{- | You set the webhook secret when you create your app. This verifies that the
+webhook is really coming from GH.
+
+Env variable: GITHUB_WEBHOOK_SECRET
+-}
+newtype GithubWebhookSecret = GithubWebhookSecret Text
+
+{- | Get the app identifier—an integer—from your app page after you create your
+app. This isn't actually a secret, but it is something easier to configure at
+runtime.
+
+Env variable: GITHUB_APP_IDENTIFIER
+-}
+newtype GithubAppIdentifier = GithubAppIdentifier Text
+
+{- | Before each request to our app, we want to instantiate an Octokit client. Doing so requires that we construct a JWT.
+* HTTP://jwt.io/introduction/
+
+We have to also sign that JWT with our private key, so GitHub can be sure that
+ a) it came from us
+ b) it hasn't been altered by a malicious third party
+-}
+data AuthPayload = AuthPayload
+    { -- | The time that this JWT was issued, _i.e._ now.
+      authPayloadIssuedAt :: !POSIXTime
+
+      {- | How long is the JWT good for (in seconds)?
+      Let's say it can be used for 10 minutes before it needs to be refreshed.
+      TODO: we don't actually cache this token, we regenerate a new one every time!
+      -}
+    , authPayloadExpiry   :: !POSIXTime
+
+      -- | Your GitHub App's identifier number, so GitHub knows who issued the
+      -- JWT, and know what permissions this token has.
+    , authPayloadId       :: !GithubAppIdentifier
+    }
+
+newtype JwtToken = JwtToken
+    { unJwtToken :: Text
+    }
+
+authPayloadToJwt :: AuthPayload -> Jwt.JWTClaimsSet
+authPayloadToJwt AuthPayload{..} = Jwt.def
+    { Jwt.iat = Jwt.numericDate authPayloadIssuedAt
+    , Jwt.exp = Jwt.numericDate authPayloadExpiry
```

Comment:
You might know already but the `jwt` library doesn't check for expiration of the JWT. As it says [here](https://bitbucket.org/ssaasen/haskell-jwt/src/6b65841ff72c1065d501e1e311000f2b5c45fac9/src/Web/JWT.hs?at=master&fileviewer=file-view-default)
```
* There is currently no verification of time related information
   ('exp', 'nbf', 'iat').
```
I know because I was working with a friend on [this PR](https://bitbucket.org/ssaasen/haskell-jwt/pull-requests/11/exp-check/diff) to implement it but it was declined for some reason. We might consider pinging the author or using `jose-jwt`.

Reply:
@sphaso You mean that `decode` function from `jwt` haskell doesn't perform verification of `exp` field during decoding?

Reply:
@ChShersh yes. I'm not sure if that responsibility should be on Hintman or Github however...

Reply:
@sphaso Good to know :+1: I think using `jwt` is fine for now, in our case `expiry` always will be valid.